src/OrleansSQLUtils/CreateOrleansTables_PostgreSql.sql

```diff
@@ -859,3 +859,213 @@ VALUES
+	
+    BEGIN
+
+	-- Grain state is not null, so the state must have been read from the storage before.
+	-- Let's try to update it.
+	--
+	-- When Orleans is running in normal, non-split state, there will
+	-- be only one grain with the given ID and type combination only. This
+	-- grain saves states mostly serially if Orleans guarantees are upheld. Even
+	-- if not, the updates should work correctly due to version number.
+	--
+	-- In split brain situations there can be a situation where there are two or more
+	-- grains with the given ID and type combination. When they try to INSERT
+	-- concurrently, the table needs to be locked pessimistically before one of
+	-- the grains gets @GrainStateVersion = 1 in return and the other grains will fail
+	-- to update storage. The following arrangement is made to reduce locking in normal operation.
+	--
+	-- If the version number explicitly returned is still the same, Orleans interprets it so the update did not succeed
+	-- and throws an InconsistentStateException.
+	--
+	-- See further information at http://dotnet.github.io/orleans/Getting-Started-With-Orleans/Grain-Persistence.
+	IF _GrainStateVersion IS NOT NULL
+	THEN
+		UPDATE Storage
+		SET
+			PayloadBinary = _PayloadBinary,
+			PayloadJson = _PayloadJson,
+			PayloadXml = _PayloadXml,
+			ModifiedOn = (now() at time zone 'utc'),
+            Version = Version + 1
+       
+		WHERE
+			GrainIdHash = _GrainIdHash AND _GrainIdHash IS NOT NULL
+			AND GrainTypeHash = _GrainTypeHash AND _GrainTypeHash IS NOT NULL
+			AND GrainIdN0 = _GrainIdN0 AND _GrainIdN0 IS NOT NULL
+			AND GrainIdN1 = _GrainIdN1 AND _GrainIdN1 IS NOT NULL
+			AND GrainTypeString = _GrainTypeString AND _GrainTypeString IS NOT NULL
+			AND ((_GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString IS NOT NULL AND GrainIdExtensionString = _GrainIdExtensionString) OR _GrainIdExtensionString IS NULL AND GrainIdExtensionString IS NULL)
+			AND ServiceId = _ServiceId AND _ServiceId IS NOT NULL
+			AND Version IS NOT NULL AND Version = _GrainStateVersion AND _GrainStateVersion IS NOT NULL;
+            
+		GET DIAGNOSTICS RowCountVar = ROW_COUNT;
+        IF RowCountVar > 0
+        THEN
+			_newGrainStateVersion := _GrainStateVersion + 1;
+        END IF;            
+            
+            
+	END IF;
+
+	-- The grain state has not been read. The following locks rather pessimistically
+	-- to ensure only on INSERT succeeds.
+    IF _GrainStateVersion IS NULL
+	THEN
+		INSERT INTO Storage
+		(
+			GrainIdHash,
+			GrainIdN0,
+			GrainIdN1,
+			GrainTypeHash,
+			GrainTypeString,
+			GrainIdExtensionString,
+			ServiceId,
+			PayloadBinary,
+			PayloadJson,
+			PayloadXml,
+			ModifiedOn,
+			Version
+		)
+		SELECT  
+			_GrainIdHash,
+			_GrainIdN0,
+			_GrainIdN1,
+			_GrainTypeHash,
+			_GrainTypeString,
+			_GrainIdExtensionString,
+			_ServiceId,
+			_PayloadBinary,
+			_PayloadJson,
+			_PayloadXml,
+           (now() at time zone 'utc'),
+			1
+		WHERE NOT EXISTS
```

Comment:
There could be a smarter way to do this in PostgreSQL, but maybe not an issue for this PR.